chips/rp2040/src/pio.rs

```diff
@@ -1156,68 +1372,83 @@ impl Pio {
         }
     }
 
-    /// Immediately execute an instruction on a state machine.
-    pub fn sm_exec(&self, sm_number: SMNumber, instr: u32) {
-        self.registers.sm[sm_number as usize]
-            .instr
-            .modify(SMx_INSTR::INSTR.val(instr));
-    }
-
-    /// Write a word of data to a state machine’s TX FIFO.
-    pub fn sm_put(&self, sm_number: SMNumber, data: u32) {
-        self.registers.txf[sm_number as usize].set(data);
-    }
-
-    /// Wait until a state machine's TX FIFO is empty, then write a word of data to it.
-    pub fn sm_put_blocking(&self, sm_number: SMNumber, data: u32) {
-        while self.registers.fstat.read(FSTAT::TXFULL0) != 0 {}
-        self.registers.txf[sm_number as usize].set(data);
+    /// Handle interrupts
+    pub fn handle_interrupt(&self) {
+        let ints = &self.registers.irq0_ints;
+        for (sm, irq) in self.sms.iter().zip([
+            IRQ0_INTS::SM0_TXNFULL,
+            IRQ0_INTS::SM1_TXNFULL,
+            IRQ0_INTS::SM2_TXNFULL,
+            IRQ0_INTS::SM3_TXNFULL,
+        ]) {
+            if ints.is_set(irq) {
+                sm.handle_tx_interrupt();
+            }
+        }
+        for (sm, irq) in self.sms.iter().zip([
+            IRQ0_INTS::SM0_RXNEMPTY,
+            IRQ0_INTS::SM1_RXNEMPTY,
+            IRQ0_INTS::SM2_RXNEMPTY,
+            IRQ0_INTS::SM3_RXNEMPTY,
+        ]) {
+            if ints.is_set(irq) {
+                sm.handle_rx_interrupt();
+            }
+        }
     }
 
-    /// Restart a state machine.
-    pub fn restart_sm(&self, sm_number: SMNumber) {
-        match sm_number {
-            SMNumber::SM0 => self.set_registers.ctrl.modify(CTRL::SM0_RESTART::SET),
-            SMNumber::SM1 => self.set_registers.ctrl.modify(CTRL::SM1_RESTART::SET),
-            SMNumber::SM2 => self.set_registers.ctrl.modify(CTRL::SM2_RESTART::SET),
-            SMNumber::SM3 => self.set_registers.ctrl.modify(CTRL::SM3_RESTART::SET),
+    /// Adds a program to PIO.
+    /// Call this with add_program(include_bytes!("path_to_file")).
+    pub fn add_program(&self, origin: Option<u8>, program: &[u8]) -> Result<u8, ProgramError> {
```

Comment:
```suggestion
    /// Adds a program to PIO, optionally relocated to a specific `origin` address.
    ///
    /// # Example
    /// 
    /// ```
    /// add_program(None, include_bytes!("path_to_file"))
    /// ```
    pub fn add_program(&self, origin: Option<u8>, program: &[u8]) -> Result<u8, ProgramError> {
```

Reply:
Adapt the other `add_program*` and `try_add_program*` method comments as well? (And of course confirm my interpretation of origin is correct.